zkwrapper/zkwrapper.cc

```diff
@@ -0,0 +1,534 @@
+	if (!zh) {
+		LOG(ERROR) << CLASS_NAME <<  "zk init failed!";
+		error_code = -999;
+	}
+	init = 1;
+	if (root_path.size() != 0) {
+		bool root_exists;
+		if (!exists(root_path, root_exists, error_code)){
+			LOG(ERROR) << CLASS_NAME <<  "Failed to check if root directory " << root << " exists " << error_code;
+			init = -1;
+			return;
+		}
+		if (!root_exists) {
+			if (!recursive_create(root_path, EMPTY_VECTOR, error_code)) {
+				LOG(ERROR) << CLASS_NAME <<  "Failed to create root directory " << root << " with error " << error_code;
+				init = -1;
+				return;
+			}
+		}
+	}
+	root = root_path;
+}
+
+std::string ZKWrapper::prepend_zk_root(const std::string& path) const {
+	if (root.size() == 0) {
+		return path;
+	}
+	if (path == "/") {
+		return root;
+	}
+	return root + path;
+}
+
+std::string ZKWrapper::removeZKRoot(const std::string& path) const {
+	return path.substr(root.size());
+}
+
+/* Wrapper Implementation of Zookeeper Functions */
+
+bool ZKWrapper::create(const std::string &path,
+		const std::vector <std::uint8_t> &data,
+		int &error_code,
+		bool ephemeral) const {
+	if (!init) {
+		LOG(ERROR) << CLASS_NAME <<  "Attempt to create before init!";
+		error_code = -999;
+		return false;
+	}
+	auto real_path = prepend_zk_root(path);
+	LOG(INFO) << CLASS_NAME <<  "creating ZNode at " << real_path;
+	int flag = (ephemeral) ? ZOO_EPHEMERAL : 0;
+	int rc = zoo_create(zh,
+			real_path.c_str(),
+			reinterpret_cast<const char *>(data.data()),
+			data.size(),
+			&ZOO_OPEN_ACL_UNSAFE,
+			flag,
+			nullptr,
+			0);
+	error_code = rc;
+	if (!rc)
+		return true;
+    LOG(ERROR) << CLASS_NAME <<  "Failed to create ZNode at " << real_path;
+    print_error(error_code);
+	return false;
+}
+
+// TODO: Modify this
+bool ZKWrapper::create_sequential(const std::string &path,
+		const std::vector <std::uint8_t> &data,
+		std::string &new_path,
+		bool ephemeral,
+		int &error_code) const {
+
+	LOG(INFO) << CLASS_NAME <<  "Starting sequential for " << path;
+	if (!init) {
+		LOG(ERROR) << CLASS_NAME <<  "Attempt to create sequential before init!";
+		return false;
+	}
+	int flag = ZOO_SEQUENCE;
+	if (ephemeral) {
+		flag = flag | ZOO_EPHEMERAL;
+	}
+	LOG(INFO) << CLASS_NAME <<  "Attempting to generate new path" << new_path;
+	LOG(INFO) << CLASS_NAME <<  "creating seq ZNode at " << prepend_zk_root(path);
+
+	int len = prepend_zk_root(path).size();
+	new_path.resize(MAX_PATH_LEN);
+	int rc = zoo_create(zh,
+			prepend_zk_root(path).c_str(),
+			reinterpret_cast<const char *>(data.data()),
+			data.size(),
+			&ZOO_OPEN_ACL_UNSAFE,
+			flag,
+			reinterpret_cast<char *>(&new_path[0]),
+			MAX_PATH_LEN);
+	error_code = rc;
+	if (rc) { // Z_OK is 0, so if we receive anything else fail
+		LOG(ERROR) << CLASS_NAME <<  "Create for " << prepend_zk_root(path) << " failed " << rc;
+		print_error(error_code);
+		return false;
+	}
+	int i = 0;
+	LOG(INFO) << CLASS_NAME <<  "NEW path is " << new_path;
+	new_path.resize(len+10);
```

Comment:
The default sequential value is a 10 digit string appended to the end. But yah, this should be a named constant